ceremonies/meetup-validation/src/tests.rs

```diff
@@ -0,0 +1,228 @@
+	let participants: Vec<usize> = vec![0, 1, 2, 3, 4];
+	let num_attestations: Vec<usize> = vec![2, 0, 3, 2, 0];
+	assert_eq!(
+		group_participants_by_num_attestations(&participants, &num_attestations),
+		vec![(0, vec![1, 4]), (2, vec![0, 3]), (3, vec![2])]
+	);
+}
+#[test]
+fn group_participants_by_num_incoming_and_outgoing_attestations_works() {
+	let participants: Vec<usize> = vec![0, 1, 2, 3, 4];
+	let participant_attestations: Vec<Vec<usize>> =
+		vec![vec![1, 2, 3], vec![3], vec![0, 1, 3], vec![1, 2], vec![0, 1, 2]];
+	let expected: (Vec<(usize, Vec<usize>)>, Vec<(usize, Vec<usize>)>) = (
+		vec![(1, vec![1]), (2, vec![3]), (3, vec![0, 2, 4])],
+		vec![(0, vec![4]), (2, vec![0]), (3, vec![2, 3]), (4, vec![1])],
+	);
+	assert_eq!(
+		group_participants_by_num_incoming_and_outgoing_attestations(
+			&participants,
+			&participant_attestations
+		),
+		expected
+	);
+}
+
+#[test]
+fn filter_attestations_works() {
+	let participants: Vec<usize> = vec![0, 2, 4];
+	let participant_attestations: Vec<Vec<usize>> =
+		vec![vec![2], vec![], vec![0], vec![2], vec![0, 2]];
+	let expected: Vec<Vec<usize>> = vec![vec![2], vec![], vec![0], vec![2], vec![0, 2]];
+	assert_eq!(filter_attestations(&participants, &participant_attestations), expected);
+}
+
+#[test]
+fn get_excluded_participants_no_vote_works() {
+	let participants: Vec<usize> = vec![0, 1, 2, 3, 4];
+	let participant_votes: Vec<u32> = vec![0, 1, 2, 2, 0];
+	let excluded_participants: Vec<usize> = vec![0, 4];
+	assert_eq!(
+		get_excluded_participants_no_vote(&participants, &participant_votes)
+			.iter()
+			.map(|p| p.0)
+			.collect::<Vec<usize>>(),
+		excluded_participants
+	);
+}
+
+#[test]
+fn get_excluded_participants_wrong_vote_works() {
+	let participants: Vec<usize> = vec![0, 1, 2, 3, 4];
+	let participant_votes: Vec<u32> = vec![3, 1, 2, 2, 5];
+	let excluded_participants: Vec<usize> = vec![0, 1, 4];
+	assert_eq!(
+		get_excluded_participants_wrong_vote(&participants, &participant_votes, 2)
+			.iter()
+			.map(|p| p.0)
+			.collect::<Vec<usize>>(),
+		excluded_participants
+	);
+}
+
+#[test]
+fn get_excluded_participants_num_attestations_works() {
+	let participants: Vec<usize> = vec![0, 1, 2, 3, 4];
+	let participant_attestations: Vec<Vec<usize>> =
+		vec![vec![1, 2, 4], vec![1], vec![0, 1, 4], vec![0, 1, 2, 4], vec![0, 1, 2]];
+	let excluded_participants: Vec<(usize, ExclusionReason)> = vec![
+		(3, ExclusionReason::TooFewIncomingAttestations),
+		(1, ExclusionReason::TooFewOutgoingAttestations),
+	];
+	assert_eq!(
+		get_excluded_participants_num_attestations(
+			&participants,
+			&participant_attestations,
+			|n| n - 1
+		),
+		excluded_participants
+	);
+}
+
+#[test]
+fn find_majority_vote_works() {
+	let participants: Vec<usize> = vec![0, 1, 2, 3, 4];
+	let participant_votes: Vec<u32> = vec![1, 1, 2, 3, 1];
+	assert_eq!(find_majority_vote(&participants, &participant_votes), Ok((1u32, 3u32)));
+}
+
+#[test]
+fn get_updated_participants_works_case_0() {
```

Comment:
your numbered cases only differ in inputs and expected outputs. Please refactor this to a parametric test `#[rstest(...`. This will make it a lot easier to include more test cases

Reply:
and I'd like to see many more cases:
* assignment size from 3..18
* attendance size: n-0, n-1.....down to zero
* attestment size n-1 down to zero
* all real cases we had so far

Reply:
about using rstest i have some concerns. i built it now and i just want to give you a feeling of how it looks:
```
#[rstest(participants, participant_votes, participant_attestations, participant_judgements,
// Scenario: Everyone attests everyone
case(vec![0, 1, 2],
	vec![3, 3, 3],
	vec![vec![1, 2], vec![0, 2], vec![0, 1]],
	ParticipantJudgements { legit: participants.clone(), excluded: vec![] }
),
// Scenario: One participant did not vote
case(vec![0, 1, 2, 3],
	vec![0, 3, 3, 3],
	vec![vec![1, 2, 3],vec![0, 2, 3], vec![0, 1, 3], vec![0, 1, 2]],
	ParticipantJudgements {
		legit: vec![1, 2, 3],
		excluded: vec![ExcludedParticipant { index: 0, reason: ExclusionReason::NoVote }],
	}
),
// Scenario: One participant did not vote like majority
case(vec![0, 1, 2, 3],
	vec![1, 3, 3, 3],
	vec![vec![1, 2, 3], vec![0, 2, 3], vec![0, 1, 3], vec![0, 1, 2]],
	ParticipantJudgements {
		legit: vec![1, 2, 3],
		excluded: vec![ExcludedParticipant { index: 0, reason: ExclusionReason::WrongVote }],
	}
),
// Scenario: 5 participants
// 0 has a broken phone and can only be attested by 1 and 2 and attests nobody
// we expect that everybody except for 0 gets a reward
case(vec![0, 1, 2, 3, 4],
vec![5, 5, 5, 5, 5],
	vec![vec![], vec![0, 2, 3, 4], vec![0, 1, 3, 4], vec![1, 2, 4], vec![1, 2, 3]],
	ParticipantJudgements {
		legit: vec![1, 2, 3, 4],
		excluded: vec![ExcludedParticipant { index: 0, reason: ExclusionReason::TooFewOutgoingAttestations }],
	}
),
// Scenario: 5 participants
// 0 has a broken phone and can only be attested by 1 and 2 and attests nobody
// we expect that everybody except for 0 gets a reward
case(vec![0, 1, 2, 3, 4],
	vec![5, 5, 5, 5, 5],
	vec![vec![1, 2, 3, 4], vec![0, 2, 3, 4], vec![0, 1, 3, 4], vec![1, 2, 4], vec![1, 2, 3]],
	ParticipantJudgements {
		legit: vec![1, 2, 3, 4],
		excluded: vec![ExcludedParticipant { index: 0, reason: ExclusionReason::TooFewIncomingAttestations }],
	}
)
)]
```

considering that we do not have code highlighting and auto format in the rstest code, i am a bit worried readbility and maintainability. @clangenb what do you think? do you know of any plugins that give us code highlighting and auto format for rstest?

Reply:
No, I don't, but formatting should be supported in attribute type proc macros.

Also, I would recommend to give your test cases names (not only documentation) with: `case::scenario5`

Maybe coud abuse the `fixures` attribute to get something like:

```
case::scenario1(test_cases(1))
fn my_test(test_case: TestCase) {...}
```

where the `test_cases(i: usize)` is a fixture that simply returns the ith element or our test cases.

Reply:
this would be possible, but also does not look very handy, because you would have to manually add a `#case` for every testcase. i also looked at `#values` but i guess it does not work very well because cannot pass a vec of testcases in there. so after some thought i came to the conclusion that this fee0634780b1888b7f405b4f1d6ef1e6cf4b468b might be the best solution. what do you think?

Reply:
I think it is very important to have a good dev experience here and it may not be overly complicated to read: So I suggest the following:

1. use rstest to vary two parameters only: number of assignees and number attendees (optionally, add a scenario-specific parameter 
2. assume fully connected and fully reciprocal adjacency matrix and then - based on the scenario - remove attestations selectively. Your current solution is way too verbose and error-prone because it's impossible to read and grasp a 10x10 adjacency matrix IMO  
3. write one rstest per scenario ("n_attendees_do_not_attest_anyone", "attendee_is_not_attested_by_n_others")